Initial commit of host metrics virtual memory scraper for windows

james-bebbington opened the pull request:
**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector/issues/847

**Description:**
Added virtual memory scraper to the hostmetricsreceiver, currently supporting Windows only, which scrapes pagefile usage and paging rate (major only) metrics via performance counters.

I will add Unix virtual memory metrics in a followup PR, as this Scraper will unfortunately need to diverge almost entirely between Windows & other OSs.

**Metrics:**

![image](https://user-images.githubusercontent.com/568630/82300743-9c37d580-99fa-11ea-8ff5-b1fbb1e69c88.png)
![image](https://user-images.githubusercontent.com/568630/82300775-a528a700-99fa-11ea-86a5-34a3073cb03d.png)

jrcamp commented on receiver/hostmetricsreceiver/internal/scraper/virtualmemoryscraper/virtualmemory_constants.go:
```diff
@@ -0,0 +1,56 @@
+func createMetricPageFileUsedDescriptor() pdata.MetricDescriptor {
+	descriptor := pdata.NewMetricDescriptor()
+	descriptor.InitEmpty()
+	descriptor.SetName("host/pagefile/percent_used")
```
What do you think about this? Would make it in line with potential cpu/memory utilization metrics.

```suggestion
	descriptor.SetName("host/pagefile/utilization")
```

james-bebbington replied:
👍 I wasn't sure if "utilization" would make sense for all metrics where we calculate percentages, but I can't think of any cases where it doesn't so I'll switch it over

james-bebbington replied:
Note I changed approach here so am generating a "usage" instead of "utilization" stat now. I'll come back and add derived "utilization" stats for all relevant metrics in a later PR.

jrcamp commented on receiver/hostmetricsreceiver/internal/scraper/virtualmemoryscraper/virtualmemory_constants.go:
```diff
@@ -0,0 +1,56 @@
+func createMetricPagingPerSecondDescriptor() pdata.MetricDescriptor {
+	descriptor := pdata.NewMetricDescriptor()
+	descriptor.InitEmpty()
+	descriptor.SetName("host/memory/paging_rate")
```
Is this the one we discussed possible reversing it and reporting as a counter?

jrcamp commented on receiver/hostmetricsreceiver/internal/scraper/virtualmemoryscraper/virtualmemory_constants.go:
```diff
@@ -0,0 +1,56 @@
+func createMetricPageFileUsedDescriptor() pdata.MetricDescriptor {
+	descriptor := pdata.NewMetricDescriptor()
+	descriptor.InitEmpty()
+	descriptor.SetName("host/pagefile/percent_used")
```
Are we going to try to make swap metrics that both exist across Windows and Linux named the same? Not sure what term would be best since Windows calls it a page file and Linux calls it swap. I'd probably call it swap but I may have a Linux bias.

jrcamp commented on receiver/hostmetricsreceiver/internal/scraper/virtualmemoryscraper/virtualmemory_constants.go:
```diff
@@ -0,0 +1,56 @@
+	descriptor := pdata.NewMetricDescriptor()
+	descriptor.InitEmpty()
+	descriptor.SetName("host/pagefile/percent_used")
+	descriptor.SetDescription("Pagefile usage (Percent).")
```
Could we call https://docs.microsoft.com/en-us/windows/win32/api/psapi/nf-psapi-enumpagefiles instead and get the absolute values and handle utilization the same way we're going to for cpu and memory?

james-bebbington replied:
👍 - I'll look into this

james-bebbington commented:
@bogdandrutu @jrcamp 

I made some pretty significant changes as per Jay's comments if you could take another look:

1. Changed terminology from "page file" -> "swap" for better consistency across Windows/Linux
2. Calculate page file usage using a syscall instead of performance counter
3. Derive paging (swap) usage from paging per second performance counter instead of reporting per second values

james-bebbington commented on receiver/hostmetricsreceiver/internal/scraper/virtualmemoryscraper/virtualmemory_scraper_windows.go:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	s.cumulativePageReads += (pageReadsPerSecValues[0].Value * durationSinceLastScraped)
+	s.cumulativePageWrites += (pageWritesPerSecValues[0].Value * durationSinceLastScraped)
```
One issue with doing this calculation is that we are now reporting float values which looks & feels a little strange (see screenshot in PR description). Do you think it would be better to convert to an integer here?

jrcamp replied:
With the underlying precision being at 1s anyway I don't think you'd lose much to convert them to integers.

james-bebbington commented on receiver/hostmetricsreceiver/internal/scraper/virtualmemoryscraper/pagefile_windows.go:
```diff
@@ -0,0 +1,74 @@
+var (
+	modPsapi           = windows.NewLazySystemDLL("psapi.dll")
+	procEnumPageFilesW = modPsapi.NewProc("EnumPageFilesW")
+)
```
Ideally everything in this file will get upstreamed to `gopsutil`. I'll create a PR in that repo as well and update here if that gets merged.

james-bebbington commented:
Regarding coverage, is it possible to override the new "codecov/patch" gate in some cases?

The problem here is the generated cover report is only based on Linux, and this PR pretty much exclusively adds code & tests for Windows, with a few lines of scaffolding for Linux. It would seem a bit overkill to hack this PR to increase coverage for that (see the codecode diff).

NB: Coverage of the virtualmemoryscraper package on Windows is: 
```shell
> go-acc .
ok      go.opentelemetry.io/collector/receiver/hostmetricsreceiver/internal/scraper/virtualmemoryscraper        0.392s  coverage: 95.3
```

jrcamp commented on receiver/hostmetricsreceiver/internal/scraper/virtualmemoryscraper/pagefile_windows.go:
```diff
@@ -0,0 +1,76 @@
+	*pageFiles = append(*pageFiles, pfData)
+
+	// return true to continue enumerating page files
+	ret := new(bool)
+	*ret = true
+	return ret
```
What about:

```suggestion
	ret := true
	return &ret
```

james-bebbington commented on receiver/hostmetricsreceiver/internal/scraper/virtualmemoryscraper/virtualmemory_scraper_others.go:
```diff
@@ -0,0 +1,164 @@
+	idps := metric.Int64DataPoints()
+	idps.Resize(1)
+	initializePageFaultDataPoint(idps.At(0), startTime, minorTypeLabelValue, int64(swap.PgFault))
+	// TODO add swap.PgMajFault once available in gopsutil
```
https://github.com/shirou/gopsutil/pull/882

james-bebbington commented on receiver/hostmetricsreceiver/internal/scraper/virtualmemoryscraper/virtualmemory_scraper_windows.go:
```diff
@@ -0,0 +1,193 @@
+
+// Initialize
+func (s *scraper) Initialize(_ context.Context) error {
+	s.startTime = pdata.TimestampUnixNano(uint64(time.Now().UnixNano()))
```
Note since we are deriving the counter from per second values, the counter essentially starts from when we take the first reading, thus why I set this to the current time here. I believe that's correct?

james-bebbington commented:
I did one more pass to this PR adding support for Linux metrics (as a second commit). Note the comment above about coverage isn't relevant anymore since I made this change.

@jrcamp - If you could pls take a look at the Linux changes, in particular:

- I'm not 100% confident about where I'm getting the swap usage data from (although this seems to match what collectd does)
- I changed the metric names/formats slightly from the original docs:
  - Combined `host/swap/paging` & `host/memory/paging` into `host/swap/paging` with `major`/`minor` labels respectively
  - Renamed `host/memory/page_faults` to `host/swap/page_faults` (I feel like it could be named either, but like the cleanliness of having all the virtual memory metrics labelled under swap ... but happy to be convinced otherwise)
- I haven't added the `host/memory/pages` metric. Not too sure how important this one is, but it doesn't appear to be currently exposed by gopsutil

jrcamp commented:
> * I'm not 100% confident about where I'm getting the swap usage data from (although this seems to match what collectd does)

Looks correct to me based on [kernel docs](https://github.com/torvalds/linux/blob/481ed297d900af0ce395f6ca8975903b76a5a59e/Documentation/filesystems/proc.rst#meminfo).

Actually their SwapFree description is confusing. It makes it sound like something different but looking at a Linux VM it appears to be what it sounds like from the name (unused swap space).

The rest sounds good!

jrcamp commented:
@bogdandrutu ptal

bogdandrutu commented on receiver/hostmetricsreceiver/internal/scraper/loadscraper/load_scraper_windows_test.go:
```diff
@@ -39,7 +39,7 @@ func TestStartSampling(t *testing.T) {
 	// that will ensure a positive value is returned
 	assert.IsType(t, &pdh.PerfCounter{}, samplerInstance.processorQueueLengthCounter)
-	samplerInstance.processorQueueLengthCounter = pdh.NewMockPerfCounter(100)
+	samplerInstance.processorQueueLengthCounter = pdh.NewMockPerfCounter(100.0)
```
In the future better to send changes unrelated to the main PR separately.
